### DISCUSSION TITLE
Add multiple fields to register dataFetcher

### DISCUSSION THREAD

**Discussion by k631583871** [graphql-java/graphql-java#2998] [Ideas]
graphql.schema.idl.TypeRuntimeWiring.Builder#dataFetcher Added support for multi field registration

**Comment by bbakerman** [comment]
Can you please expand on the ask?

Do you want a way to more easily provide data fetchers other than one by one?

Have you checked out `graphql.schema.idl.WiringFactory` and specifically `graphql.schema.idl.WiringFactory#providesDataFetcher` and `graphql.schema.idl.WiringFactory#getDataFetcher`

**Reply by k631583871** [reply]
This API can't meet my needs. I need to bind multiple fields. When querying bound fields, I only call DataFetcher#get once

For implementation https://github.com/spring-projects/spring-graphql/issues/519 This function

**Comment by bbakerman** [comment]
You have told us what does not work - but you really haven't told us what you want to do clearly.

I can guess that English is not you first language but given the information I have and the linked issue, I am not closer to understanding what you want to achive.

**Reply by k631583871** [reply]
Expected to add methods in graphql.schema.idl.TypeRuntimeWiring.Builder
```java
public Builder dataFetcher(List<String> fieldNames, DataFetcher dataFetcher) {

      return this;
}
```
The following TDOD functions are what I want to achieve.
```java
    public static void main(String[] args) {
        String schema = "type Query{\n" +
                "    hello: Bean\n" +
                "}\n" +
                "\n" +
                "type Bean{\n" +
                "    hello0:String\n" +
                "    hello1:String\n" +
                "    hello2:String\n" +
                "    hello3:String\n" +
                "    hello4:String\n" +
                "}";

        SchemaParser schemaParser = new SchemaParser();
        TypeDefinitionRegistry typeDefinitionRegistry = schemaParser.parse(schema);


        RuntimeWiring runtimeWiring = RuntimeWiring.newRuntimeWiring()
                .type("Query", builder -> builder.dataFetcher("hello", new StaticDataFetcher("hello0")))

                .type("Bean", builder -> builder.dataFetcher("hello0", new StaticDataFetcher("hello0")))
                .type("Bean", builder -> builder.dataFetcher("hello1", new StaticDataFetcher("hello1")))
                // TODO start
                .type("Bean", builder -> builder.dataFetcher(Arrays.<String>asList("hello3","hello4"), environment -> {
                    // Trigger only once
                    System.out.println("name:" + environment.getField().getName());
                    Map<String, String> map = new HashMap<>();
                    map.put("hello3", "hello3");
                    map.put("hello4", "hello4");
                    return map;
                }))
                // TODO end
                .build();

        SchemaGenerator schemaGenerator = new SchemaGenerator();
        GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);

        GraphQL build = GraphQL.newGraphQL(graphQLSchema).build();
        ExecutionResult executionResult = build.execute("{hello {hello0,hello1,hello2,hello3,hello4}}");

        System.out.println(executionResult.getData().toString());
        // TODO Expected output
        // Prints: {hello={hello0=hello0, hello1=hello1, hello2=null, hello3=hello3, hello4=hello4}}
    }
```

**Reply by bbakerman** [reply]
Ahh I understand - you wanted a sort of batch DataFetcher that can get N attributes at a time and have them fit into the object.

Yeah we fundamentally don't have that.  We have followed the reference graphql-js implementation such that each field has a data fetcher and it returns one value.

**Reply by k631583871** [reply]
Do you consider adding this function?
I think this function can reduce the number of queries to the actual database in some special cases.

**Reply by bbakerman** [reply]
Typically you would use sub selection read ahead

See https://www.graphql-java.com/blog/deep-dive-data-fetcher-results

**Reply by k631583871** [reply]
I understand that graphql is more suitable as a data gateway. It is necessary to use this function because it will involve migrating data sources in the application iteration process.
I plan to use Reactive and cache to obtain data, and use simple code to aggregate a large number of fields to reduce development difficulty.

**Reply by bbakerman** [reply]
Did you have a look at sub selection read ahead.  This allows you to see what sub fields have been asked for and then allows you to prefetch based on the asked for fields

**Reply by k631583871** [reply]
Thank you for your example. I have seen the example you mentioned. That example can meet my needs.
I hope to get a simpler api, which is more convenient for developing functions.
